src/main/java/com/google/sps/servlets/ImageServlet.java

```diff
@@ -0,0 +1,160 @@
+
+    response.getWriter().println(uploadUrl);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PrintWriter writer = response.getWriter();
+    BlobKey blobKey = getBlobKey(request, "image");
+    if (blobKey == null) {
+      response.sendRedirect("/index.html");
+      return;
+    }
+
+    byte[] blobBytes = getBlobBytes(blobKey);
+    List<EntityAnnotation> annotations = getImageLabels(blobBytes);
+    writer.println("Annotations size: " + annotations.size());
+    writer.println("Blob Key: " + blobKey);
+    writer.println("Number of bytes: " + blobBytes.length);
+   
+    // Create an array of labels for printing the appropriate JSON 
+    List<String> imageLabels = new ArrayList<>();;
+    for (int i = 0; i < imageLabels.size() && i < MAX_NUM_KEYWORDS; i++) {
+      imageLabels.add(annotations.get(i).getDescription());
+    }
+    Gson gson = new Gson();
+    String json = gson.toJson(imageLabels);
+    writer.println(json);
+    response.sendRedirect("/main.html");
+  }
+
+  /**
+   * Returns the BlobKey that points to the file uploaded by the user, or null if the user didn't
+   * upload a file.
+   */
+  private BlobKey getBlobKey(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a BlobKey.
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+    System.out.println(blobKey.getKeyString());
+
+    // User submitted form without selecting a file, so the BlobKey is empty.
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    return blobKey;
+  }
+
+  /**
+   * Retrieves the binary data stored at the BlobKey parameter.
+   */
+  private byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // end index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // if we read fewer bytes than we requested, then we reached the end
+      if (b.length < fetchSize) {
```

Comment:
There's no need, since if the file is too small, then the loop will terminate after the first iteration (but it still makes sure to continue reading the rest of the bytes from the image).